src/Order/Order.php

```diff
@@ -0,0 +1,547 @@
+
+            $orderItem->OrderItemAddOns()->remove($addOn);
+        }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function canView($member = null)
+    {
+        return $this->extendedCan(__FUNCTION__, $member) ??
+            Permission::check(ShopPermissions::VIEW_ORDERS, 'any', $member);
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function canEdit($member = null)
+    {
+        return $this->extendedCan(__FUNCTION__, $member) ??
+            ($this->IsMutable() && Permission::check(ShopPermissions::EDIT_ORDERS, 'any', $member));
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function canDelete($member = null)
+    {
+        return false;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function canCreate($member = null, $context = [])
+    {
+        return false;
+    }
+
+    /**
+     * @param null|Member $member
+     * @param string[] $guestTokens
+     * @return bool
+     */
+    public function canViewOrderPage(?Member $member = null, array $guestTokens = []): bool
+    {
+        if ($this->IsMutable()) {
+            // No one should be able to view carts as an order
+            return false;
+        }
+
+        $extendedCan = $this->extendedCan('canViewOrderPage', $member, [
+            'guestTokens' => $guestTokens,
+        ]);
+
+        // Allow if extendedCan, admin or valid guest token
+        return $extendedCan ?? (in_array($this->GuestToken, $guestTokens, true) ||
+                Permission::checkMember($member, 'ADMIN'));
+    }
+
+    /**
+     * @throws \Exception
+     */
+    public function Lock(): void
+    {
+        if ($this->CartLocked) return;
+
+        DB::get_conn()->withTransaction(function () {
+            foreach ($this->OrderItems() as $item) {
+                if ($item->getQuantity() < 1) {
+                    $item->delete();
+                } else {
+                    $item->PurchasableLockedVersion = $item->Purchasable()->Version;
+                    $item->write();
+                }
+            }
+
+            $this->CartLocked = true;
+            $this->write();
+        }, null, false, true);
+    }
+
+    /**
+     * @return bool
+     */
+    public function Empty(): bool
+    {
+        return !$this->exists() || !boolval($this->OrderItems()->sum('Quantity'));
+    }
+
+    /**
+     * @param Payment $payment
+     * @param ServiceResponse $response
+     */
+    public function paymentCaptured(Payment $payment, ServiceResponse $response): void
+    {
+        $this->ConfirmationTime = DBDatetime::now();
+        $this->Environment = Director::get_environment_type();
+        $this->IsCart = false;
+        $this->write();
+
+        if (!$this->UnpaidTotal()->getMoney()->isPositive()) {
+            OrderConfirmationEmail::create($this)->send();
+        }
+    }
+
+    /**
+     * @param null|Payment $payment
+     * @throws \Exception
+     */
+    public function paymentCancelled(?Payment $payment): void
```

Comment:
Could use some phpdoc on some of these methods. Not sure what a lot of these are meant to do.